Batteries/Data/Vector/Monadic.lean

```diff
@@ -0,0 +1,171 @@
+/-
+Copyright (c) 2024 Shreyas Srinivas. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Shreyas Srinivas
+-/
+
+import Batteries.Data.Vector.Basic
+import Init.Data.Array.Basic
+import Init.Data.Array.Lemmas
+import Batteries.Classes.SatisfiesM
+
+/-!
+## Monadic Definitions for Vectors
+In this file we add monadic definitions for statically sized vectors.
+-/
+
+namespace Batteries
+
+namespace Vector
+
+/--
+`modifyM v i f` applies a monadic transformation `f` to `v[i]`
+-/
+def modifyM [Monad m] (v : Vector α n) (i : Nat)
+  (f : α → m α) : m (Vector α n) := do
+    return ⟨←Array.modifyM v.toArray i f,
+      by rw[←v.size_eq]; sorry⟩
+
+/--
+`modify v i f` takes a vector `v`, index `i`, and a modification function `f`
+and sets `v[i]` to `f`.
+-/
+@[inline]
+def modify (v : Vector α n) (i : Nat) (f : α → α) : Vector α n :=
+  Id.run <| modifyM v i f
+
+/--
+`forIn v acc f` iterates over `v` applying `f` on each element and the
+accumulated result. The final value of `acc` after the completion of the loop
+is returned.
+-/
+protected def forIn {α : Type u} {β : Type v} {m : Type v → Type w} [Monad m]
+  (v : Vector α n) (acc : β) (f : α → β → m (ForInStep β)) : m β := do
+    Array.forIn v.toArray acc f
+
+instance : ForIn m (Vector α n) α where
+  forIn := Vector.forIn
+
+/--
+`foldlM f init v start stop` is the monadic left fold function that folds `f`
+over `v` from left to right.
+-/
+def foldlM {α : Type u} {β : Type v} {m : Type v → Type w} [Monad m]
+  (f : β → α → m β) (init : β)
+  (v : Vector α n) (start := 0) (stop := n) : m β := do
+    Array.foldlM f init v.toArray start stop
+
+/--
+`foldrM f init v start stop` is the monadic right fold function that folds `f`
+over `v` from right to left.
+-/
+def foldrM {α : Type u} {β : Type v} {m : Type v → Type w} [Monad m]
+  (f : α → β → m β) (init : β) (v : Vector α n)
+  (start := n) (stop := 0) : m β :=
+    Array.foldrM f init v.toArray start stop
+
+/--
+`mapM f v` maps a monadic function `f : α → m β` over `v` and returns
+the resultant vector in the same monad `m`.
+-/
+def mapM {α : Type u} {β : Type v} {m : Type v → Type w} [Monad m]
+  (f : α → m β) (v : Vector α n) : m (Vector β n) := do
+    return ⟨←Array.mapM f v.toArray, by rw[←v.size_eq]; sorry⟩
```

Comment:
This `sorry` is also stuck for the same reason as the above [comment](https://github.com/leanprover-community/batteries/pull/925/files#r1720862336) because the corresponding Array API uses Satisfies. I am grateful for any help with these situations